Update hook component to use enum MCA parameter

jjhursey opened the pull request:
* `--mca ompi_display_comm`
   - `mpi_init` : Display during `MPI_Init`
   - `mpi_finalize` : Display during `MPI_Finalize`

jjhursey commented:
This was a community request to tighten up the MCA parameter before v5.

jjhursey commented:
* `mpirun` CLI option in https://github.com/openpmix/prrte/pull/711
 * The core functionality was introduced in PR #5507

jsquyres commented on ompi/mca/hook/comm_method/hook_comm_method_component.c:
```diff
@@ -132,21 +145,28 @@ static int ompi_hook_comm_method_component_register(void)
-    if( 1 == hook_comm_method ) {
+    mca_base_var_enum_create("ompi_comm_method", mca_hook_comm_method_modes, &new_enum);
+    ret = mca_base_var_register("ompi", NULL, NULL, "comm_method",
+                                "Enable comm_method behavior (1) mpi_init or (2) mpi_finalize",
```
Does this help message need to be updated to show the words and/or the 3/all option?

jjhursey replied:
Yeah, that's probably a good idea. It's part of the output from the registration, but I can clean up the main description.

jjhursey replied:
Fixed in latest push.

hjelmn commented:
Another option is the flag enumerator. That would allow it to take comma-delimited values such as "init,finalize".

jjhursey commented:
@hjelmn That sounds interesting. I like that more than `all`. Let me play around with it a bit - I'll push a new commit for review.

hjelmn commented:
Yeah, I created it because of ```btl_*_flags```. I could never remember the flags for put, get, etc. Not we can do ```--mca btl_ugni_flags send,put,get``` and the like.

hjelmn commented:
For an example:

https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/base/btl_base_frame.c#L41
https://github.com/open-mpi/ompi/blob/master/opal/mca/btl/base/btl_base_mca.c#L49

jjhursey commented:
I just pushed a new commit to the branch that uses the enum flags instead of just an enum. I also took out the '1' and '2' options since they are not really useful for this option (there are more verbose options).

```
--------------------------------------------------------------------------
An invalid value was supplied for an enum variable.

  Variable     : comm_method
  Value        : 12
  Valid values : Comma-delimited list of:  0x1:"none", 0x2:"init", 0x2:"mpi_init", 0x4:"finalize", 0x4:"mpi_finalize"
--------------------------------------------------------------------------
```

Something odd (not related to this PR) is that the `ompi_comm_method` MCA parameter is not shown in `ompi_info` - possibly because it is not registered to a specific component?

jsquyres commented:
I think you might want to remove the duplicate value.  I.e., pick `init` / `finalize` or `mpi_init` / `mpi_finalize` -- don't have both.

In this mode of MCA values, those become distinct / different values.

jjhursey commented:
I don't think I understand - they are aliases for each other
```
0x2:"init", 0x2:"mpi_init"
0x4:"finalize", 0x4:"mpi_finalize"
``
So `init` and `mpi_init` are both the value `0x2`

jjhursey commented:
I see it now. There is a check in `mca_base_var_enum.c:400` that checks to:
```
/* ensure flags are only set a single bit, doesn't conflict with itself, and
 * hasn't already been specified. */
```
Since I was using a debug build it didn't trip when I tested (and worked fine). I took out the duplicates leaving only `mpi_init` and `mpi_finalize`

jjhursey commented:
I fixed the `ompi_info` display issue and squashed everything down. This is ready to go if anyone wants to take a second look.

```
           MCA ompi base: parameter "ompi_comm_method" (current value: "none", data source: default, level: 3 user/all, type: unsigned_int)
                          Enable the communication protocol report: when MPI_INIT is invoked (using the 'mpi_init' value) and/or when MPI_FINALIZE is invoked (using the 'mpi_finalize' value).
                          Valid values: Comma-delimited list of:  0x1:"none", 0x2:"mpi_init", 0x4:"mpi_finalize"
```

jsquyres commented on ompi/mca/hook/comm_method/hook_comm_method_component.c:
```diff
@@ -113,61 +134,70 @@ static int ompi_hook_comm_method_component_register(void)
     mca_hook_comm_method_enable_mpi_finalize = false;
     (void) mca_base_component_var_register(&mca_hook_comm_method_component.hookm_version, "enable_mpi_finalize",
-                                           "Enable comm_method behavior on mpi_finalize",
+                                           "Enable the communication protocol report when MPI_FINALIZE is invoked",
```
Just curious -- why do we have `hook_comm_method_enable_mpi_init`, `hook_comm_method_enable_mpi_finalize`, _and_ `ompi_comm_method`?  They seem to be redundant.

Should we pick one set or the other?  I kinda like the `ompi_comm_method` enum flags, but have no strong preference here.

jjhursey replied:
That's a fair point. I'll take out the other two options, and leave the `ompi_comm_method`.

jjhursey replied:
Fixed in latest push.

jsquyres commented on ompi/mca/hook/comm_method/hook_comm_method_component.c:
```diff
@@ -111,63 +132,53 @@ static int ompi_hook_comm_method_component_register(void)
-    // comm_method_max
-    (void) mca_base_var_register("ompi", NULL, NULL, "comm_method_max",
+    // hook_comm_method_max
+    (void) mca_base_component_var_register(&mca_hook_comm_method_component.hookm_version, "max",
```
Did you intend these changes to be part of this commit / PR?  It changes the MCA param name from `ompi_comm_method_max` to `hook_comm_method_max`.  That isn't described in the commit message at all.

Also, there's now a disparity between `ompi_comm_method` and `hook_comm_method_*` -- why are some in the OMPI level and some in the hook framework level?

jjhursey replied:
Yeah, I meant them to be there. It seemed odd to have a bunch of `ompi_` specific controls for this. We wanted the `ompi_comm_method` to be general - not tied to a component - for ease of use/reference. The others seemed to me to be fine-tuning which seems like a component thing.

That said, since we intend to hide it behind the CLI option in https://github.com/openpmix/prrte/pull/711 we could make them all component level MCAs. Do you have a preference?

jsquyres replied:
Yeah, I'm sorry to be so nit-picky, but having fewer one-off situations as a good thing.

How about a compromise -- because the PRRTE PR CLI option is good for (effectively) `mpirun`, but not for direct launch:

1. Make the MCA param be `hook_...whatever` (i.e., properly prefixed)
2. Make an alias MCA param to it that is just `ompi_comm_method`.  That alias can have a useful help message and/or comment in the code explaining why it's not `hook_...`.

Indeed, I imagine that that should be fairly "normal" for hook components: they're OMPI-wide functionality.  So having an `ompi_FOO` MCA param seems appropriate.  But It is probably at least worth mentioning as a comment in the code, yadda yadda yadda.

Thanks!

jjhursey replied:
How about the ompi level MCA being `ompi_display_comm` to match the discussion [here](https://github.com/openpmix/prrte/pull/711#issuecomment-756268334)?

jsquyres replied:
Works for me.

jjhursey replied:
Done. Let me know what you think.

jjhursey commented:
Here is how the MCA parameters show up in ompi_info now:
```
shell$  ompi_info --all -l 9 | grep -A4 "hook comm_method\|ompi_display_comm"
    MCA hook comm_method: ---------------------------------------------------
    MCA hook comm_method: parameter "hook_comm_method_verbose" (current value: "0", data source: default, level: 9 dev/all, type: int)
    MCA hook comm_method: parameter "hook_comm_method_display" (current value: "none", data source: default, level: 3 user/all, type: unsigned_int, synonyms: ompi_display_comm)
                          Enable the communication protocol report: when MPI_INIT is invoked (using the 'mpi_init' value) and/or when MPI_FINALIZE is invoked (using the 'mpi_finalize' value).
                          Valid values: Comma-delimited list of:  0x1:"none", 0x2:"mpi_init", 0x4:"mpi_finalize"
    MCA hook comm_method: parameter "hook_comm_method_max" (current value: "12", data source: default, level: 3 user/all, type: int)
                          Number of hosts for which to print unabbreviated 2d table of comm methods.
    MCA hook comm_method: parameter "hook_comm_method_brief" (current value: "0", data source: default, level: 3 user/all, type: int)
                          Only print the comm method summary, skip the 2d table.
    MCA hook comm_method: parameter "hook_comm_method_fakefile" (current value: "", data source: default, level: 3 user/all, type: string)
                          For debugging only: read comm methods from a file
```

jjhursey commented:
Per offline discussion removed the 'none' option since not specifying the parameter is really 'none' anyway.

```
    MCA hook comm_method: parameter "hook_comm_method_display" (current value: "", data source: default, level: 3 user/all, type: unsigned_int, synonyms: ompi_display_comm)
                          Enable the communication protocol report: when MPI_INIT is invoked (using the 'mpi_init' value) and/or when MPI_FINALIZE is invoked (using the 'mpi_finalize' value).
                          Valid values: Comma-delimited list of:  0x1:"mpi_init", 0x2:"mpi_finalize"
```
